### PR TITLE
add support for filter list

### PR DESCRIPTION
**Pull request by otnielvh**

### CODE REVIEW

File: src/indexer.c [hunk 1]
```diff
@@ -154,52 +185,80 @@ RedisModuleDict * QueryIndexPredicate(RedisModuleCtx *ctx, QueryPredicate *predi
...
+        case NEQ:
+        case EQ:
+            value = RedisModule_StringPtrLen(predicate->label.value, &_s);
+            index_key = RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key, value);
+            break;
+        default:
+            break;
+        }
+
+    if (predicate->type == FILTER_LIST) {
```
Comment (danni-m) [inline src/indexer.c]: @otnielvh the parsing should not be here, `QueryPredicate` should be changed to keep multi-value, and the parsing function for the predicates `parseLabelListFromArgs` (should be renamed and return/update a predicate)

**Comment by K-Jo** [comment]
#73

File: src/indexer.c [hunk 2]
```diff
@@ -28,22 +39,44 @@ int parseLabel(RedisModuleCtx *ctx, RedisModuleString *label, Label *retLabel, c
...
+    if (token == NULL)
+        return TSDB_ERROR;
+    retLabel->key = RedisModule_CreateString(NULL, token, strlen(token));
+
+    // Extract value
+    token = strtok_r(NULL, separator, &iter_ptr);
+    if (strcmp(separator, "=(") == 0)  // for filter_list we must remove the closing parentheses
     {
-        if (token == NULL && i==0) {
+        if (token == NULL)
```
Comment (danni-m) [inline src/indexer.c]: its always better to use if (...) { }

File: src/indexer.c [hunk 3]
```diff
@@ -153,53 +199,74 @@ RedisModuleDict * QueryIndexPredicate(RedisModuleCtx *ctx, QueryPredicate *predi
...
+        }
+
+    if (predicate->type == FILTER_LIST) {
+
+        for (int i=0; i < predicate->label.listCount; i++){
+            const char* singleValue = RedisModule_StringPtrLen(predicate->label.valueList[i], &_s);
+            index_key = RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key, singleValue);
+            currentLeaf = RedisModule_DictGet(labelsIndex, index_key, &nokey);
+            if (nokey == 0) {
+                _deepCopyDict(ctx, localResult, currentLeaf);
```
Comment (danni-m) [inline src/indexer.c]: _deepCopyDict should be renamed to union

File: src/indexer.h [hunk 4]
```diff
@@ -11,7 +11,13 @@
 
 typedef struct {
     RedisModuleString *key;
-    RedisModuleString *value;
+    union {
```
Comment (danni-m) [inline src/indexer.h]: Im not a big fan of this, whats the benefits of using a union over just adding `valueList` and `listCount`?

Comment (danni-m) [reply src/indexer.h]: I think this complexity should reside in QueryPredicate and Label should be used only for the key's label storage.

Comment (otnielvh) [reply src/indexer.h]: So Label will contain a kay and value, and the QueryPredicate will contain the valueList and listCount?

File: docs/commands.md [hunk 5]
```diff
@@ -211,7 +211,12 @@ TS.MRANGE fromTimestamp toTimestamp [AGGREGATION aggregationType bucketSizeSecon
 
 * fromTimestamp - Start timestamp for range query
 * toTimestamp - End timestamp for range query
-* filters - Set of key-pair fitlers (`k=v`, `k!=v,` `k=` contains a key, `k!=` doesn't contain a key)
+* filters - Set of key-pair fitlers. The supported filters are:
+  * `k=v` label equals value
+  * `k!=v` label doesn't equal value
+  * `k=` key contains the label `k`
```
Comment (yaeltzirulnikov) [inline docs/commands.md]: @danni-m if I'm not mistaken, this is the opposite
'k=' means key doesn't contain label and
'k!=' means key does contain label, no?

File: src/indexer.c [hunk 6]
```diff
@@ -15,6 +15,17 @@ RedisModuleDict *labelsIndex;
 #define KV_PREFIX "__index_%s=%s"
 #define K_PREFIX "__key_index_%s"
 
+void _deepCopyDict(RedisModuleCtx *ctx, RedisModuleDict *dest, RedisModuleDict *src) {
+    /*
+     * Copy all elements from src to dest
```
Comment (yaeltzirulnikov) [inline src/indexer.c]: might be good to add to the comment / indicate in the function name that the function only copies the keys and not the values

File: docs/commands.md [hunk 7]
```diff
@@ -350,7 +360,12 @@ Get all the keys matching the filter list.
 TS.QUERYINDEX fitler...
```
Comment (yaeltzirulnikov) [inline docs/commands.md]: might want to fix 'fitler' while you're at it:)

File: src/indexer.c [hunk 8]
```diff
@@ -153,53 +199,74 @@ RedisModuleDict * QueryIndexPredicate(RedisModuleCtx *ctx, QueryPredicate *predi
...
+            break;
+        default:
+            break;
+        }
+
+    if (predicate->type == FILTER_LIST) {
+
+        for (int i=0; i < predicate->label.listCount; i++){
+            const char* singleValue = RedisModule_StringPtrLen(predicate->label.valueList[i], &_s);
+            index_key = RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key, singleValue);
```
Comment (yaeltzirulnikov) [inline src/indexer.c]: I think it's more efficient not to allocate new RedisModuleString here, just work on the two char* strings and then use 'RedisModule_DictGetC'.

File: docs/commands.md (hunk 5 above)
Comment (otnielvh) [reply docs/commands.md]: Not confusing at all....

File: src/indexer.c (hunk 6 above)
Comment (otnielvh) [reply src/indexer.c]: Given Danni's comment, the name is changed to _union, so it works the same as the _intersection etc.

File: docs/commands.md (hunk 5 above)
Comment (otnielvh) [reply docs/commands.md]: Anyway, I didn't change the content just the formatting (i.e. new lines), so we can open a separate bug about this, and decide what the correct behaviour is.

**Comment by danni-m** [comment]
@otnielvh please rebase.

File: src/module.c [hunk 9]
```diff
@@ -262,24 +262,30 @@ void ReplyWithAggValue(RedisModuleCtx *ctx, timestamp_t last_agg_timestamp, Aggr
 
 int parseLabelListFromArgs(RedisModuleCtx *ctx, RedisModuleString **argv, int start, int query_count,
         QueryPredicate *queries) {
+
```
Comment (danni-m) [inline src/module.c]: redundant white space

File: src/indexer.c [hunk 10]
```diff
@@ -173,30 +225,82 @@ RedisModuleDict * QueryIndexPredicate(RedisModuleCtx *ctx, QueryPredicate *predi
...
+            default:
+                index_key = NULL;
         }
-        return prevResults;
-    } else if (predicate->type == EQ || predicate->type == CONTAINS) {
-        return localResult;
-    } else {
-        return prevResults;
+
+        if (predicate->type == FILTER_LIST) {
```
Comment (danni-m) [inline src/indexer.c]: this function should not have *any* parsing code in it.

File: src/indexer.c [hunk 11]
```diff
@@ -173,30 +225,82 @@ RedisModuleDict * QueryIndexPredicate(RedisModuleCtx *ctx, QueryPredicate *predi
...
+        const char *value;
 
-    if (prevResults != NULL) {
-        if (predicate->type == EQ || predicate->type == CONTAINS) {
-            _intersect(ctx, prevResults, localResult);
-        } else  if (predicate->type == NCONTAINS) {
-            _difference(ctx, prevResults, localResult);
-        } else if (predicate->type == NEQ){
-            _difference(ctx, prevResults, localResult);
+        switch (predicate->type) {
```
Comment (danni-m) [inline src/indexer.c]: this is too complex, the easier think to do is:
 RedisModuleDict *localResult = RedisModule_CreateDict(ctx);
for item in predicate-> valuesList:
   _union(localResult, GetDictFromIndex(item))

File: src/indexer.c [hunk 12]
```diff
@@ -21,30 +21,57 @@ void IndexInit() {
...
     char *labelstr = RedisModule_PoolAlloc(ctx, _s + 1);
     labelstr[_s] = '\0';
     strncpy(labelstr, labelRaw, _s);
+
+    // Extract key
     token = strtok_r(labelstr, separator, &iter_ptr);
-    for (int i=0; i<2; i++)
+    if (token == NULL)
+        return TSDB_ERROR;
+    retQuery->key = RedisModule_CreateString(NULL, token, strlen(token));
```
[the remark this answers is not in the document]
Comment (otnielvh) [reply src/indexer.c]: Ops, you're right. Good catch

File: src/indexer.c [hunk 13]
```diff
@@ -134,29 +172,35 @@ void _difference(RedisModuleCtx *ctx, RedisModuleDict *left, RedisModuleDict *ri
...
-        const char *key = RedisModule_StringPtrLen(predicate->label.key, &_s);
-        const char *value = RedisModule_StringPtrLen(predicate->label.value, &_s);
-        index_key = RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key, value);
-    }
     int nokey;
-    currentLeaf = RedisModule_DictGet(labelsIndex, index_key, &nokey);
-    if (nokey) {
 
-        currentLeaf = NULL;
+    if (predicate->valuesList == 0) {
```
Comment (danni-m) [inline src/indexer.c]:
```suggestion
        if (predicate->type == NCONTAINS || predicate->type == CONTAINS) {	
```

File: src/indexer.c [hunk 14]
```diff
@@ -21,30 +21,56 @@ void IndexInit() {
     labelsIndex = RedisModule_CreateDict(NULL);
 }
 
-int parseLabel(RedisModuleCtx *ctx, RedisModuleString *label, Label *retLabel, const char *separator) {
+int parseLabel(RedisModuleCtx *ctx, RedisModuleString *label, QueryPredicate *retQuery, const char *separator) {
     char *token;
```
Comment (otnielvh) [inline src/indexer.c]: parsePredicate

File: src/indexer.c [hunk 15]
```diff
@@ -134,29 +172,35 @@ void _difference(RedisModuleCtx *ctx, RedisModuleDict *left, RedisModuleDict *ri
...
+        RedisModuleDict *singleEntryLeaf;
+
+        for (int i = 0; i < predicate->valueListCnt; i++) {
+            value = RedisModule_StringPtrLen(predicate->valuesList[i], &_s);
+            index_key = RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key, value);
+            singleEntryLeaf = RedisModule_DictGet(labelsIndex, index_key, &nokey);
+            if (currentLeaf == NULL) {
+                currentLeaf = singleEntryLeaf;
+            } else if (singleEntryLeaf != NULL){
+                _union(ctx, currentLeaf, singleEntryLeaf);
```
Comment (danni-m) [inline src/indexer.c]: you are mutating `currentLeaf` which is the actual index.

Comment (danni-m) [reply src/indexer.c]: lets open a ticket to optimize the 2nd copy in `QueryIndexPredicate`

File: src/indexer.c [hunk 16]
```diff
@@ -190,12 +234,14 @@ RedisModuleDict * QueryIndexPredicate(RedisModuleCtx *ctx, QueryPredicate *predi
             _difference(ctx, prevResults, localResult);
         } else if (predicate->type == NEQ){
             _difference(ctx, prevResults, localResult);
+        } else if (predicate->type == FILTER_LIST) {
+            _union(ctx, prevResults, localResult);
```
Comment (danni-m) [inline src/indexer.c]:
```suggestion
            _intersect(ctx, prevResults, localResult);
```

File: src/indexer.h [hunk 17]
```diff
@@ -21,19 +21,22 @@ typedef enum  {
...
     // REQ,
     // NREQ
 } PredicateType;
 
 typedef struct QueryPredicate {
     PredicateType type;
-    Label label;
+    RedisModuleString *key;
+    RedisModuleString **valuesList;
+    int valueListCnt;
```
Comment (danni-m) [inline src/indexer.h]: not critical
```suggestion
    int valueListCount;
```

**Pull request merged**